dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/OpenAPIVersionConfigImpl.java

```diff
@@ -0,0 +1,282 @@
+/*******************************************************************************
+ * Copyright (c) 2024 IBM Corporation and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package io.openliberty.microprofile.openapi20.internal;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.eclipse.microprofile.openapi.models.OpenAPI;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Modified;
+
+import com.ibm.websphere.ras.Tr;
+import com.ibm.websphere.ras.TraceComponent;
+import com.ibm.ws.ffdc.annotation.FFDCIgnore;
+import com.ibm.ws.kernel.productinfo.ProductInfo;
+
+import io.openliberty.microprofile.openapi20.internal.services.OpenAPIVersionConfig;
+import io.openliberty.microprofile.openapi20.internal.utils.MessageConstants;
+
+/**
+ * Base implementation of OpenAPIVersionConfig which supports configuring version 3.0.x of OpenAPI
+ */
+public class OpenAPIVersionConfigImpl implements OpenAPIVersionConfig {
+    private static final TraceComponent tc = Tr.register(OpenAPIVersionConfigImpl.class);
+
+    private static final String VERSION_KEY = "openAPIVersion";
+    protected static final OpenAPIVersion VERSION_30 = new OpenAPIVersion(3, 0, -1);
+    protected static final OpenAPIVersion VERSION_303 = new OpenAPIVersion(3, 0, 3);
+
+    protected volatile OpenAPIVersion configuredVersion = defaultVersion();
+    protected String configuredVersionString = null;
+
+    @Activate
+    @Modified
+    protected void activate(Map<?, ?> properties) {
+        if (!ProductInfo.getBetaEdition()) {
+            // Do not read config if we're not the beta edition, to avoid potential for errors while parsing
+            return;
+        }
+
+        String versionString = (String) properties.get(VERSION_KEY);
+
+        if (Objects.equals(versionString, configuredVersionString)) {
+            // Our config has not changed, return immediately to avoid potentially logging warnings
+            if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
+                Tr.event(this, tc, "Config unchanged", versionString);
```

Comment:
An edge case. Lets say the customer sets it to null. On activate configuredVersion will be set to defaultVersion(). Then on modified null != defaultVersion.

The only effect is we'll get the wrong event message "no OpenAPI version set" rather than "Config unchanged".

Reply:
Not quite.

If the user sets it to `null`, we'll set `configuredVersion` to the `defaultVersion()` and `configuredVersionString` to `null`.

The equality check is done against `configuredVersionString`.